cdap-app-fabric/src/main/java/co/cask/cdap/metadata/FieldLineageAdmin.java

```diff
@@ -0,0 +1,308 @@
+      }
+    }
+
+    Stream<Map.Entry<ProgramId, Long>> sortedByLastExecutedTime
+            = programIdToLastExecutedTime.entrySet().stream()
+                                          .sorted(Collections.reverseOrder(Map.Entry.comparingByValue()));
+
+    List<ProgramInfo> programInfos = new ArrayList<>();
+    sortedByLastExecutedTime.forEachOrdered(programIdLongEntry
+            -> programInfos.add(new ProgramInfo(programIdLongEntry.getKey(), programIdLongEntry.getValue())));
+
+    return programInfos;
+  }
+
+  /**
+   * Computes list of {@link FieldOperationInfo} from the given operations.
+   * Returned list contains the operations sorted in topological order i.e. each operation
+   * in the list is guaranteed to occur before any other operation which reads its outputs.
+   * @param operations set of operation to convert to FieldOperationInfo instances
+   * @return list of FieldOperationInfo sorted topologically
+   */
+  private List<FieldOperationInfo> computeFieldOperationInfo(Set<Operation> operations) {
+    Set<String> writeOperations = new HashSet<>();
+    Map<String, Operation> operationsMap = new HashMap<>();
+    for (Operation operation : operations) {
+      operationsMap.put(operation.getName(), operation);
+      if (operation.getType().equals(OperationType.WRITE)) {
+        writeOperations.add(operation.getName());
+      }
+    }
+
+    LinkedHashSet<String> orderedOperations = new LinkedHashSet<>();
+    for (String write : writeOperations) {
+      WriteOperation writeOperation = (WriteOperation) operationsMap.get(write);
+      List<InputField> inputs = writeOperation.getInputs();
+      Set<String> visitedOperationNames = new LinkedHashSet<>();
+      visitedOperationNames.add(writeOperation.getName());
+      for (InputField input : inputs) {
+        Operation operation = operationsMap.get(input.getOrigin());
+        if (operation != null) {
+          // since we return subset of operations, it is possible that some of the origins are
+          // not in a subset so we need to do null check
+          computeFieldOperationInfoHelper(operationsMap, operationsMap.get(input.getOrigin()), visitedOperationNames);
+        }
+      }
+      orderedOperations.addAll(visitedOperationNames);
+    }
+
+    List<FieldOperationInfo> fieldOperationInfos = new ArrayList<>();
+    for (String operationName : orderedOperations) {
+      Operation operation = operationsMap.get(operationName);
+      if (operation != null) {
+        // since we return subset of operations, it is possible that some of the origins are
+        // not in a subset so we need to do null check
+        fieldOperationInfos.add(convertToFieldOperationInfo(operationsMap.get(operationName)));
+      }
+    }
+
+    Collections.reverse(fieldOperationInfos);
+    return fieldOperationInfos;
+  }
+
+  private void computeFieldOperationInfoHelper(Map<String, Operation> operationsMap, Operation currentOperation,
```

Comment:
what does this method do? does it compute the subset of the operations that are reachable from the write operations? It would help you name your methods after what they do ("helper" is not very descriptive) and add a javadoc

Reply:
restructured the code and removed this method.